core/unix/signal.c

```diff
@@ -6693,6 +6700,141 @@ execute_native_handler(dcontext_t *dcontext, int sig, sigframe_rt_t *our_frame)
+    // O: Each Other thread wakes up, does sig_detach which reinstates the app
+    //    signal stack (if available) and the app's blocked signal set, lets the
+    //    Detacher thread know that its done detaching and resumes native
+    //    execution. Note that at this point, DR's sigact config is still
+    //    installed (including main_signal_handler). When the Other thread
+    //    return from the suspend signal, signals get unblocked automatically
+    //    and may now be delivered.
+    // <execute_native_handler_using_cur_frame may help here>
+    // D: When all threads are done detaching, the Detacher thread cleans up
+    //    each thread's drcontext for real, and also reinstates the app's
+    //    sigact in signal_thread_exit (which removes main_signal_handler).
+    //
+    // Therefore, if we arrive here and find doing_detach to be true and
+    // dcontext to be NULL, we can be certain that the current thread has
+    // completed sig_detach. This means the interrupted context stack (sc->XSP)
+    // must be the app's own regular stack.
+    //
+    // Also, if the thread has a signal stack at all for any signal, we will
+    // be on it (because DR sets SA_ONSTACK for all signals, and that config is
+    // still in effect), regardless of whether the signal stack was registered
+    // for sig. If the signal stack was actually not registered for sig, then we
+    // return true (indicating that DR should continue trying to deliver the
+    // signal).
+    if (dcontext != NULL || !doing_detach)
+        return false;
+
+    bool has_sigstack = false;
+#ifdef HAVE_SIGALTSTACK
+    stack_t app_sigstack;
+    IF_DEBUG(int rc =)
+    sigaltstack_syscall(NULL, &app_sigstack);
+    ASSERT(rc == 0);
+    if (APP_HAS_SIGSTACK_EX(app_sigstack)) {
+        has_sigstack = true;
+        // If the app has a sigstack, we must be on it.
+        ASSERT(app_sigstack.ss_sp < (void *)xsp &&
+               (void *)xsp < app_sigstack.ss_sp + app_sigstack.ss_size);
+        // In release build, we bail on this optimization and try to deliver the
+        // signal.
+        return false;
+    }
+#endif
+
+    kernel_sigaction_t sigact_struct;
+    d_r_read_lock(&detached_sigact_lock);
+    memcpy(&sigact_struct, &detached_sigact[sig], sizeof(sigact_struct));
+    d_r_read_unlock(&detached_sigact_lock);
+
+#ifdef HAVE_SIGALTSTACK
+    // Satisfy "will never be NULL" complain from the compiler.
+    kernel_sigaction_t *sigact = &sigact_struct;
+    if (has_sigstack && !USE_APP_SIGSTACK_EX(app_sigstack, sigact)) {
+        // We are on the sigstack but we don't want to use it for signal delivery.
+        return false;
+    }
+#endif
+
+    if (sigact_struct.handler == (handler_t)SIG_IGN ||
+        sigact_struct.handler == (handler_t)SIG_DFL) {
+        // We handle logs, messages and asserts related to these cases in
+        // execute_native_handler() later.
+        return false;
+    }
+
```

Comment:
Where is the check for the app wanting an RT frame, which we discussed offline?  We can't use the current RT frame if the app expects a non-RT frame.

Reply:
I'm not very familiar with the RT distinction. From get_app_frame_size in signal.c I understand IS_RT_FOR_APP is what I need?

Reply:
Yes.

Reply:
That's done already then.